packages/drivers/odsp-driver/src/fetchSnapshot.ts

```diff
@@ -58,13 +64,71 @@ export async function fetchSnapshot(
     );
 }
 
+export async function redeemSharingLink(
+    odspResolvedUrl: IOdspResolvedUrl,
+    storageTokenFetcher: (options: TokenFetchOptions, name: string) => Promise<string | null>,
+) {
+    await getWithRetryForTokenRefresh(async (tokenFetchOptions) => {
+        assert(odspResolvedUrl.sharingLinkToRedeem !== undefined, "Share link should be present");
+        const storageToken = await storageTokenFetcher(tokenFetchOptions, "TreesLatest");
+        const shareId = encodeShareUrl(odspResolvedUrl.sharingLinkToRedeem);
+        assert(shareId !== undefined, "ShareId should be present for share link");
+        const redeemUrl = `${odspResolvedUrl.siteUrl}/_api/v2.0/shares/${shareId}`;
+        const { url, headers } = getUrlAndHeadersWithAuth(redeemUrl, storageToken);
+        headers.prefer = "redeemSharingLink";
+        return fetchAndParseAsJSONHelper(url, { headers });
+    });
+}
+
+export async function fetchSnapshotWithRedeem(
+    odspResolvedUrl: IOdspResolvedUrl,
+    storageTokenFetcher: (options: TokenFetchOptions, name: string) => Promise<string | null>,
+    snapshotOptions: ISnapshotOptions | undefined,
+    logger: ITelemetryLogger,
+    snapshotDownloader: (url: string, fetchOptions: {[index: string]: any}) => Promise<IOdspResponse<IOdspSnapshot>>,
+    putInCache: (valueWithEpoch: IVersionedValueWithEpoch) => Promise<void>,
+): Promise<ISnapshotCacheValue> {
+    let odspSnapshot: ISnapshotCacheValue;
+    try {
+        odspSnapshot = await fetchLatestSnapshotCore(
+            odspResolvedUrl,
+            storageTokenFetcher,
+            snapshotOptions,
+            logger,
+            snapshotDownloader,
+            putInCache,
+        );
+    } catch(error) {
+        if (isRedeemSharingLinkError(odspResolvedUrl, error)) {
+            await redeemSharingLink(odspResolvedUrl, storageTokenFetcher);
+            logger.sendErrorEvent({
+                eventName: "TreeLatest_SecondCall",
+                errorType: error.errorType,
+            });
+            odspSnapshot = await fetchLatestSnapshotCore(
+                odspResolvedUrl,
+                storageTokenFetcher,
+                snapshotOptions,
+                logger,
+                snapshotDownloader,
+                putInCache,
+                false,
+            );
+        } else {
+            throw error;
+        }
+    }
+    return odspSnapshot;
+}
+
 export async function fetchLatestSnapshotCore(
     odspResolvedUrl: IOdspResolvedUrl,
     storageTokenFetcher: (options: TokenFetchOptions, name: string) => Promise<string | null>,
     snapshotOptions: ISnapshotOptions | undefined,
     logger: ITelemetryLogger,
     snapshotDownloader: (url: string, fetchOptions: {[index: string]: any}) => Promise<IOdspResponse<IOdspSnapshot>>,
     putInCache: (valueWithEpoch: IVersionedValueWithEpoch) => Promise<void>,
+    redeemShareLinkIfPresent: boolean = true,
```

Comment:
I feel like it's better not to add this new argument, but rather in places where redeeming is not needed, make a copy of odspResolvedUrl and clear odspResolvedUrl.sharingLinkToRedeem

Reply:
It would also make other flows more coherent, for example if other calls are added in the future to isRedeemSharingLinkError(), it would stay correct (vs. today it's possible that odspResolvedUrl.sharingLinkToRedeem was not used but we still use it as a pivot)